compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs

```diff
@@ -165,84 +165,143 @@ checkTemplate module_ template = do
-        throwIfDifferent "signatories" (extractFuncFromFuncThis . tplSignatories <$> template) $
-            EUpgradeError $ TemplateChangedSignatories $ NM.name $ present template
+        whenDifferent "signatories" (extractFuncFromFuncThis . tplSignatories) template $
+            warnWithContext $ WTemplateChangedSignatories $ NM.name $ present template
     withContext (ContextTemplate (present module_) (present template) TPObservers) $
-        throwIfDifferent "observers" (extractFuncFromFuncThis . tplObservers <$> template) $
-            EUpgradeError $ TemplateChangedObservers $ NM.name $ present template
+        whenDifferent "observers" (extractFuncFromFuncThis . tplObservers) template $
+            warnWithContext $ WTemplateChangedObservers $ NM.name $ present template
     withContext (ContextTemplate (present module_) (present template) TPAgreement) $
-        throwIfDifferent "agreement" (extractFuncFromFuncThis . tplAgreement <$> template) $
-            EUpgradeError $ TemplateChangedAgreement $ NM.name $ present template
+        whenDifferent "agreement" (extractFuncFromFuncThis . tplAgreement) template $
+            warnWithContext $ WTemplateChangedAgreement $ NM.name $ present template
+
+    withContext (ContextTemplate (present module_) (present template) TPKey) $ do
+        case fmap tplKey template of
+           Upgrading { past = Nothing, present = Nothing } -> do
+               pure ()
+           Upgrading { past = Just pastKey, present = Just presentKey } -> do
+               let tplKey = Upgrading pastKey presentKey
+
+               -- Key type musn't change
+               let keyTypesMatch = foldU alphaType (fmap tplKeyType tplKey)
+               unless keyTypesMatch $
+                   throwWithContext (EUpgradeError (TemplateChangedKeyType (NM.name (present template))))
+
+               -- But expression for computing it may
+               whenDifferent "key expression"
+                   (extractFuncFromFuncThis . tplKeyBody) tplKey
+                   (warnWithContext $ WTemplateChangedKeyExpression $ NM.name $ present template)
+               whenDifferent "key maintainers"
+                   (extractFuncFromTyAppNil . tplKeyMaintainers) tplKey
+                   (warnWithContext $ WTemplateChangedKeyMaintainers $ NM.name $ present template)
+           Upgrading { past = Just pastKey, present = Nothing } ->
+               throwWithContext $ EUpgradeError $ TemplateRemovedKey (NM.name (present template)) pastKey
+           Upgrading { past = Nothing, present = Just presentKey } ->
+               warnWithContext $ WTemplateAddedKeyDefinition (NM.name (present template)) presentKey
+
     -- TODO: Check that return type of a choice is compatible
     pure ()
     where
-        extractFuncFromFuncThis :: Expr -> Maybe ExprValName
-        extractFuncFromFuncThis expr
+        extractFuncFromFuncThis :: Expr -> Module -> Either String Expr
```

Comment:
Each extract function takes an expression, extracts a relevant ExprValName, and performs lookups necessary to get the actual definition.

Reply:
Consider adding that as a comment.

Reply:
Yeah I often find that writing out the haskell that these types of functions will take makes reading much easier